### ISSUE TITLE
Inquiry about training convergence

### ISSUE THREAD

**Issue by bindita** [Tetratrio/stylegan2_pytorch#7]
Thanks @Tetratrio for this code. After how many iterations can we expect to see a decent result if we train from scratch?

Here's a finetuning experiment with G_loss and D_loss plots. I don't see any visual improvement after 10,000 iterations and the D_loss also seems to be too weak. The results are ok but lack the high fidelity as in the original paper. However, I am training on 2300 images, so the small dataset size could be a problem.

![loss](https://user-images.githubusercontent.com/43757801/89166913-96e52400-d52f-11ea-80fb-87a3174ebb93.png)

Also, do you know why stylegan doesn't use any VGG / perceptual loss? I generally use this loss for high fidelity in other methods, but seems strange they got such good results without this loss.

**Comment by adriansahlman** [comment]
Hey, glad you found it useful!

I did not have the hardware to train with the full 1024x1024 resolution in the paper but got quite similar results for 512x512. What I did notice when downloading the weights from the official implementation and using them to generate images was that there is a lot of terrible samples and the ones you generally see in the paper are curated.

The best way to get better quality images (although with less fidelity) is to use the truncation trick on the disentangled latents (``truncation_psi`` in the code and on the command line).

When training for 512x512 I noticed that I needed slightly less iterations than what was specified in the paper, probably because of the lower parameter count in my model. I found that after a while it stopped improving where running it even longer could even sometimes make it worse.

I am unsure about the loss as it can be hard to evaluate based on that its an adversarial objective. I think the gradient norm can be interesting to look at as well although as when I was playing around with different settings it would behave differently. But GANs in general are quite mysterious, I know in the paper for Big GAN they tried to analyze behavior of collapse in training [here](https://arxiv.org/pdf/1809.11096.pdf#figure.caption.4).

For perceptual loss you actually need a supervised target for your generator output. As GAN's use random input, it would be hard to find the target for each respective input, especially as the same noise as input during training will have different outputs  as training progresses (for example, some input may create a female in the beginning of training while later on it creates a male instead). The only way I can think of using perceptual loss is to calculate the statistics for the entire dataset and then try to keep the output of the generator from deviating from those statistics too much. But the loss would have to be very gentle as to not create a mode collapse where the generator learns to create only one image that has those perceptual statistics.

Perceptual loss works great with super resolution models as these models take an image as input and actually has the same image (although in higher resolution) as target.

I would have loved to experiment more with these things but I only had time during my vacation around christmas so I did not get to play around with it too much.

When it comes to the dataset size, I do think 2,300 images is quite small compared to the 50,000 used in the paper. I am not familiar with finetuning of GANs so maybe it is possible to use so few images when starting out with pretrained weights.

There should be some metrics in the repository that you can use that may help determine when training is done (perceptual path length and FID). The precision/recall metric from the paper is missing from this repo Im afraid.

As I am feeling a bit rusty on the GANs since I have not worked on them for half a year any thoughts I give here might not be correct so take them with a grain of salt!

Good luck!

**Comment by bindita** [comment]
Thanks a lot for the insights!
